File: src/memory_management_technique/LabReport4.java
package memory_management_technique;
import java.util.Arrays;
import java.util.Scanner;
public class LabReport4 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of blocks: ");
        int numBlocks = scanner.nextInt();
        int[] blockSize = new int[numBlocks];
        int[] blockSizeCopy = new int[numBlocks];
        int totalMemory = 0;
        int totalMemoryUsed = 0;

        for (int i = 0; i < numBlocks; i++) {
            System.out.print("Enter size of block " + (i + 1) + ": ");
            blockSize[i] = scanner.nextInt();
            blockSizeCopy[i] = blockSize[i];
            totalMemory += blockSize[i];
            totalMemoryUsed += blockSize[i];
        }

        System.out.print("Enter the number of processes: ");
        int numProcesses = scanner.nextInt();
        int[] processSize = new int[numProcesses];
        int totalProcessSizes = 0;
        int totalProcessSizesUsed = 0;

        for (int i = 0; i < numProcesses; i++) {
            System.out.print("Enter memory required for process " + (i + 1) + ": ");
            processSize[i] = scanner.nextInt();
            totalProcessSizes += processSize[i];
        }

        int[] blockAllocation = new int[numBlocks];
        Arrays.fill(blockAllocation, -1);

        for (int i = 0; i < numProcesses; i++) {
            for (int j = 0; j < numBlocks; j++) {
                if (blockSize[j] >= processSize[i]) {
                    blockAllocation[j] = i;
                    blockSize[j] -= processSize[i];
                    totalMemoryUsed -= processSize[i];
                    totalProcessSizesUsed += processSize[i];
                    break;
                }
            }
        }

        System.out.println("\nProcesses\tProcesses size\tBlocks\tBlocks size\tAllocated\tInt. Frag.");
        for (int i = 0; i < numProcesses; i++) {
            System.out.print("\t"+(i + 1) + "   \t\t\t" + processSize[i] +"\t\t\t");
            int allocatedBlock = -1;
            for (int j = 0; j < numBlocks; j++) {
                if (blockAllocation[j] == i) {
                    allocatedBlock = j;
                    break;
                }
            }
            if (allocatedBlock != -1) {
                System.out.print((allocatedBlock + 1) + "\t\t" + blockSizeCopy[allocatedBlock] +
                        "    \t\tYES\t\t\t " + blockSize[allocatedBlock]);
            } else {
                System.out.print("-\t\t-\t\tNO\t\t-");
            }
            System.out.println();
        }

        System.out.println("\nTotal Memory: " + totalMemory);
        System.out.println("Total Memory Used/Allocated: " + (totalMemory - totalMemoryUsed));
        System.out.println("Total Process Sizes: " + totalProcessSizes);
        System.out.println("Total Process Sizes Used/Allocated: " + totalProcessSizesUsed);
    }
}

File: src/memory_management_technique/kotlin/LabReport_4.kt
package memory_management_technique.kotlin
import java.util.*
fun main() {
    val scanner = Scanner(System.`in`)
    fun inputArray(size: Int, prompt: String) = IntArray(size) {
        print("$prompt ${it + 1}: ")
        scanner.nextInt()
    }
    print("Enter the number of Blocks– ")
    val numBlocks = scanner.nextInt()
    val blockSize = inputArray(numBlocks, "Block")
    val blockSizeCopy = blockSize.copyOf()
    print("Enter the number of processes– ")
    val numProcesses = scanner.nextInt()
    val processSize = inputArray(numProcesses, "Enter memory required for process")
    val blockAllocation = IntArray(numBlocks) { -1 }
    for (i in 0 until numProcesses) {
        val j = blockSize.indexOfFirst { it >= processSize[i] }
        if (j != -1) {
            blockAllocation[j] = i
            blockSize[j] -= processSize[i]
        }
    }
    println("Processes\tProcesses size\tBlocks\tBlocks size\tAllocated\tInt. Frag.")
    for (i in 0 until numProcesses) {
        val allocatedBlock = blockAllocation.indexOfFirst { it == i }
        with(processSize[i]) {
            print("\t${i + 1}\t\t\t$this\t\t\t")
            if (allocatedBlock != -1) { print("${allocatedBlock + 1}\t\t\t${blockSizeCopy[allocatedBlock]}" +
                    "\t\t\tYES\t\t\t${blockSize[allocatedBlock]}")
            } else {
                print("-\t\t\t-\t\t\tNO\t\t\t-")
            }
            println()
        }
    }
    val tbu = blockAllocation.count { it != -1 }
    val tpa = numProcesses - blockAllocation.count { it == -1 }
    val totalRAM = blockSizeCopy.sum()
    val totalRAMUsed = blockSize.sum()
    val totalRAMLeft = totalRAM - totalRAMUsed
    val tpsa = processSize.sum()
    println("\nTotal RAM: $totalRAM")
    println("Total RAM Used/Allocated: $totalRAMUsed")
    println("Total RAM Left: $totalRAMLeft")
    println("Total Process Size Allocated: $tpsa")
    println("\nTotal Blocks of Memory Used/Allocated: $tbu")
    println("Total Processes Allocated: $tpa")
}
